File: sn-mnndvl.py
################## Imports ##################

import numpy as np
import pandas as pd
import os
import sys
import pickle
# from MemoryNetwork import MemoryNeuralNetwork
from mnn_torch import MemoryNeuralNetwork
import matplotlib.pyplot as plt
import torch
# Seeds

np.random.seed(0)

################## DNN ##################
neeta=1e-3
neeta_dash=5e-4
lipschitz_constant = 1.2
epochs = 70
pred_list = []

mnn = MemoryNeuralNetwork(10, 50, 3, neeta=neeta, neeta_dash=neeta_dash,
                          lipschitz_norm=lipschitz_constant, spectral_norm=True)

################# MAIN ################

# load
inputcsvpath = '/home/pramuk/IISC/sn-mnn-auv-nav/inputs/input_data.csv'
targetcsvpath = '/home/pramuk/IISC/sn-mnn-auv-nav/inputs/target_var.csv'

if os.path.isfile(inputcsvpath) and os.path.isfile(targetcsvpath):
    input_data_df = pd.read_csv(inputcsvpath, index_col=0)
    input_data = input_data_df.to_numpy()
    target_var_df = pd.read_csv(targetcsvpath, index_col=0)
    Z = target_var_df.to_numpy()

else:
    print('File not present')
    exit()

error_sum = 0.0
error_list = np.zeros(epochs)
rmse_list = np.zeros(epochs)
unstable_flag = False
rmse_sum = 0.0
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
try:
    for _ in range(0, 2000):
        mnn.feedforward(np.zeros(10))
        mnn.backprop(np.zeros(3))
    # for i in range(10):
    #     mnn.feedforward(input_data[1, :])
    #     mnn.backprop(Z[1, :])
    for epoch in range(0, epochs):
        pred_list = []
        if (epoch != 0):
            print("Training for epoch %2d finished with average rmse loss of %.5f\n" % (
                epoch-1, rmse_sum/len(input_data)))
            error_list[epoch] = error_sum / len(input_data)
            rmse_list[epoch] = rmse_sum/len(input_data)
        error_sum = 0
        rmse_sum = 0
        print("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")

        for i in range(1, len(input_data)):

            if (mnn.squared_error > 1e30):

                unstable_flag = True
                # double break
                i = sys.maxint
                epoch = sys.maxint

            pred = mnn.feedforward(input_data[i-1, :])
            mnn.backprop(Z[i-1, :])

            # print(pred,Z[i-1,:])
            if (epoch > 45):
                pred_list.append(pred)

            error_sum += mnn.squared_error
            rmse_sum += mnn.rmse

            print("Training for epoch %2d, progress %5.2f%% with squared loss: %.5f with rmse: %.5f rmse x: %.5f rmse y: %.5f rmse z: %.5f" % (
                epoch, (i/len(input_data)) * 100, mnn.squared_error, mnn.rmse, mnn.rmse_x, mnn.rmse_y, mnn.rmse_z), end="\r")
        print(pred_list)

    print("=====================================================================================================")

except Exception as e:
    # print(e.__doc__)
    print(e)

finally:
    save_path = '/home/pramuk/IISC/sn-mnn-auv-nav/trained_models_mnn/sn-mnn.obj'
    if not unstable_flag:
        print("Done! saving model as " + save_path + " ...")

        dfpred = pd.DataFrame(pred_list)
        dfpred.to_csv(
            '/home/pramuk/IISC/sn-mnn-auv-nav/predictions.csv', index=False)

        filename = open(save_path, "wb")
        pickle.dump(mnn, filename)
        filename.close()

        plt.figure(0)
        plt.plot(error_list[1:])
        plt.xlabel("Epochs")
        plt.ylabel("Squared Loss")
        plt.title("Squared Loss vs Epochs")
        plt.grid(True)

        plt.figure(1)
        plt.plot(rmse_list[1:])
        plt.xlabel("Epochs")
        plt.ylabel("Root Mean Squared Loss")
        plt.title("Root mean Squared Loss vs Epochs")
        plt.grid(True)

        plt.figure(2)
        plt.plot(dfpred[0])
        plt.plot(Z[:, 0])
        plt.legend(['predicted y', 'true y'])
        plt.xlabel("Time")
        plt.ylabel("Predcition and Desired speed(x component) (m/s)")

        plt.figure(3)
        plt.plot(dfpred[1])
        plt.plot(Z[:, 1])
        plt.title("Predcition and Desired output(y component) vs Time")
        plt.legend(['predicted y', 'true y'])
        plt.xlabel("Time")
        plt.ylabel("Predcition and Desired speed(x component) (m/s)")

        plt.figure(4)
        plt.plot(dfpred[2])
        plt.plot(Z[:, 2])
        plt.title("Predcition and Desired output (z component) vs Time")
        plt.legend(['predicted z', 'true z'])
        plt.xlabel("Time")
        plt.ylabel("Predcition and Desired speed(x component) (m/s)")

        # plt.savefig("trained_models/loss_" + str(neeta) + "_" + str(neeta_dash) + "_" + str(epochs) + ".eps", format="eps", dpi=1000)
        plt.show()
    else:
        print("Network Unstable! Quitting...")

File: sn_mnndvl.m
% Import necessary libraries
%addpath('/path/to/MemoryNetwork'); % Make sure to replace with the actual path

% Set random seed
rng(10);


% Define constants and parameters
neeta = 1e-3;
neeta_dash = 5e-4;
lipschitz_constant = 1.2;
epochs = 70;

% Create MemoryNeuralNetwork object
mnn = MemoryNeuralNetwork(10, 50, 3, neeta, neeta_dash, lipschitz_constant, true);

% Load data
path = fileparts(mfilename(''));
IMU_in = load(fullfile(path, 'dataset/TrainAndValidation/IMU.mat'));
V = load(fullfile(path, 'dataset/TrainAndValidation/V.mat'));
/home/airl/auvnav/sn-mnn-auv-nav
% DVL speed to beams - pitch=20 deg eqn 4 substitution
b1 = [cos((45 + 0 * 90) * pi / 180) * sin(20 * pi / 180), ...
    sin((45 + 0 * 90) * pi / 180) * sin(20 * pi / 180), cos(20 * pi / 180)];
b2 = [cos((45 + 1 * 90) * pi / 180) * sin(20 * pi / 180), ...
    sin((45 + 1 * 90) * pi / 180) * sin(20 * pi / 180), cos(20 * pi / 180)];
b3 = [cos((45 + 2 * 90) * pi / 180) * sin(20 * pi / 180), ...
    sin((45 + 2 * 90) * pi / 180) * sin(20 * pi / 180), cos(20 * pi / 180)];
b4 = [cos((45 + 3 * 90) * pi / 180) * sin(20 * pi / 180), ...
    sin((45 + 3 * 90) * pi / 180) * sin(20 * pi / 180), cos(20 * pi / 180)];

A = [b1; b2; b3; b4]
p_inv = (A' * A) \ A';
beams = zeros(length(V), 4);
for i = 1:length(V)
    beams(i, :) = (A*(V.data(:, i)) * (1 + 0.007))'; % scale factor 0.7%
end

beams_noise = beams + (0.042^2) * randn(length(V), 4) + 0.001 * ones(length(V), 4);

T = 100; % Sampling rate IMU - 100Hz, DVL - 1Hz
X_gyro = zeros(length(IMU_in.data(1, :, 1)/T), 3);
X_acc = zeros(length(IMU_in.data(1, :, 1) / T), 3);
Y = zeros(length(IMU_in.data(1, :, 1) / T), 4);
Z = zeros(3,length(V.data(1, :) / T));

n = 1;
V = V';

for t = 1:T:length(IMU_in.data(1, :, 1))

    X_acc(n, :) = IMU_in.data(:, t, 1);
    X_gyro(n, :) = IMU_in.data(:, t, 2);
    y = beams_noise(n, :);
    Y(n, :) = y; % DVL
    z = V.data(n, :); % velocities target variable
    Z(n, :) = z;
    n = n + 1;

end

% Num of DVL samples
N = length(IMU_in.data(1, :, 1) / T);

input_data = [X_acc, X_gyro, Y];
input_data_df = array2table(input_data);
disp(length(input_data));

error_sum = 0.0;
error_list = zeros(1, epochs);
rmse_list = zeros(1, epochs);
unstable_flag = false;
rmse_sum = 0.0;

try
    for i = 1:2000
        mnn.feedforward(zeros(10, 1));
        mnn.backprop(zeros(3, 1));
    end
    for epoch = 1:epochs
        pred_list = [];
        if epoch ~= 1
            fprintf('Training for epoch %2d finished with average loss of %.5f\n', epoch - 1, error_sum / length(input_data));
            error_list(epoch) = error_sum / length(input_data);
            rmse_list(epoch) = rmse_sum

 length(input_data);
        end
        error_sum = 0;
        rmse_sum = 0;
        fprintf('++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n');

        for i = 2:length(input_data)
            if mnn.squared_error > 1e30
                unstable_flag = true;
                break;
            end
            pred = mnn.feedforward(input_data(i - 1, :)');
            mnn.backprop(Z(i - 1, :)');
            if epoch > 45
                pred_list = [pred_list; pred'];
            end
            error_sum = error_sum + mnn.squared_error;
            rmse_sum = rmse_sum + mnn.rmse;
            fprintf('Training for epoch %2d, progress %5.2f%% with squared loss: %.5f with rmse: %.5f\r', epoch, (i / length(input_data)) * 100, mnn.squared_error, mnn.rmse);
        end
    end
    fprintf('\n=====================================================================================================\n');

catch ME
    fprintf('%s', ME.message);
end

% Define RMSE function
function [rmse_ls, rmse_predicted, improv] = RMSE(true, predicted, LS)
    true = vecnorm(true, 2, 2);
    predicted = vecnorm(predicted, 2, 2);
    LS = vecnorm(LS, 2, 2);
    rmse_ls = sqrt(mean((true - LS).^2));
    rmse_predicted = sqrt(mean((true - predicted).^2));
    improv = 100 * (1 - (rmse_predicted / rmse_ls));
end

% Define MAE function
function [mae_ls, mae_predicted] = MAE(true, predicted, LS)
    true = vecnorm(true, 2, 2);
    predicted = vecnorm(predicted, 2, 2);
    LS = vecnorm(LS, 2, 2);
    mae_ls = sum(abs(LS - true)) / length(true);
    mae_predicted = sum(abs(predicted - true)) / length(true);
end

% Define NSE_R2 function
function [r2_ls, r2_predicted] = NSE_R2(true, predicted, LS)
    true = vecnorm(true, 2, 2);
    predicted = vecnorm(predicted, 2, 2);
    LS = vecnorm(LS, 2, 2);
    true_avg = mean(true);
    temp_ls = sum((LS - true).^2) / sum((true - true_avg).^2);
    r2_ls = 1 - temp_ls;
    temp_predicted = sum((predicted - true).^2) / sum((true - true_avg).^2);
    r2_predicted = 1 - temp_predicted;
end

% Define VAF function
function [r2_ls, r2_predicted] = VAF(true, predicted, LS)
    true = vecnorm(true, 2, 2);
    predicted = vecnorm(predicted, 2, 2);
    LS = vecnorm(LS, 2, 2);
    true_var = var(true);
    temp_ls = var(true - LS);
    r2_ls = (1 - temp_ls / true_var) * 100;
    temp_predicted = var(true - predicted);
    r2_predicted = (1 - temp_predicted / true_var) * 100;
end
